lib/tree.rb

```diff
@@ -1,66 +1,135 @@
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    if current_node.nil?
+      return TreeNode.new(key, value)
+    end 
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end 
+    return current_node
+  end 
+  
+  # Time Complexity: O(log(n) best case, O(n) worst case
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
+  end
+  
+  def find_helper(current_node, key)
+    if current_node.nil?
+      return nil 
+    end 
+    
+    if key > current_node.key 
+      current_node.right = find_helper(current_node.right, key)
+    elsif key < current_node.key
+      current_node.left = find_helper(current_node.left, key)
+    else
+      return current_node.value
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity:  O(n)
   def inorder
-    raise NotImplementedError
+    arr = []
+    if @root.nil?
+      return arr
+    end 
+    current_node = @root
+    return inorder_helper(current_node, arr)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current_node, arr)
+    if current_node.nil?
+      return arr
+    end 
+    inorder_helper(current_node.left, arr)
+    arr << {key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, arr)
+    return arr
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    arr = []
+    if @root.nil?
+      return arr
+    end 
+    current_node = @root
+    return preorder_helper(current_node, arr)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current_node, arr)
+    if current_node.nil?
+      return arr
+    end 
+    arr << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, arr)
+    preorder_helper(current_node.right, arr)
+    return arr
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    arr = []
+    if @root.nil?
+      return arr
+    end 
+    current_node = @root
+    return postorder_helper(current_node, arr)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
-  def height
-    raise NotImplementedError
+  
+  def postorder_helper(current_node, arr)
+    if current_node.nil?
+      return arr
+    end 
+    postorder_helper(current_node.left, arr)
+    postorder_helper(current_node.right, arr)
+    arr << { key: current_node.key, value: current_node.value }
+    return arr
   end
-
-  # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
-  def bfs
-    raise NotImplementedError
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) - stack
+  def height(current_node = @root)
```

Comment:
👍 , the space complexity is correct if the tree is unbalanced, it's O(log n) if the tree is balanced.